src/main/java/me/botsko/prism/actions/BlockAction.java

```diff
@@ -30,574 +30,534 @@
+                && !parameters.getProcessType().equals(PrismProcessType.UNDO)) {
+            Prism.debug("Block skipped because it's not allowed to be placed unless its an UNDO."
+                    + block.getType().name());
+            return new ChangeResult(ChangeResultType.SKIPPED, null);
+        }
+        // If we're not in a preview, actually apply this block
+        // Capture the block before we change it
+        final BlockState originalBlock = block.getState();
+        if (!isPreview) {
+            return handleApply(block, originalBlock, parameters, cancelIfBadPlace);
+        } else {
+
+            // Otherwise, save the state so we can cancel if needed
+            // Note: we save the original state as both old/new so we can re-use
+            // blockStateChanges
+            stateChange = new BlockStateChange(originalBlock, originalBlock);
+
+            // Preview it
+            EntityUtils.sendBlockChange(player, block.getLocation(), getBlockData());
+
+            // Send preview to shared players
+            for (final CommandSender sharedPlayer : parameters.getSharedPlayers()) {
+                if (sharedPlayer instanceof Player) {
+                    EntityUtils.sendBlockChange((Player) sharedPlayer, block.getLocation(), getBlockData());
+                }
+            }
+            return new ChangeResult(ChangeResultType.APPLIED, stateChange);
+        }
+    }
+
+    /**
+     * The BlockState object is modified by this method.
+     *
+     * @param block            Block
+     * @param originalBlock    BlockState
+     * @param parameters       QueryParameters
+     * @param cancelIfBadPlace cancelIfBadPlace
+     * @return ChangeResult.
+     */
+    private ChangeResult handleApply(final Block block, final BlockState originalBlock,
+                                     final QueryParameters parameters, final boolean cancelIfBadPlace) {
+        BlockState state = block.getState();
+        // If lily pad, check that block below is water. Be sure
+        // it's set to stationary water so the lily pad will sit
+        switch (getMaterial()) {
+            case LILY_PAD:
+                final Block below = block.getRelative(BlockFace.DOWN);
+                if (below.getType().equals(WATER) || below.getType().equals(AIR)) {
+                    below.setType(WATER);
+                } else {
+                    // Prism.debug("Lilypad skipped because no water exists below.");
+                    return new ChangeResult(ChangeResultType.SKIPPED, null);
+                }
+                break;
+            case NETHER_PORTAL: // Only way is to set the portal on fire.
+                final Block obsidian = BlockUtils.getFirstBlockOfMaterialBelow(OBSIDIAN, block.getLocation());
+                if (obsidian != null) {
+                    final Block above = obsidian.getRelative(BlockFace.UP);
+                    if (!(above.getType() == NETHER_PORTAL)) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use opposite operator instead of the logic complement operator.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5446892)